src/WaveClip.cpp

```diff
@@ -1380,51 +1321,69 @@ void WaveClip::Paste(double t0, const WaveClip* other)
-   const WaveClip* pastedClip;
+
+   t0 = std::clamp(t0, GetPlayStartTime(), GetPlayEndTime());
+
+   //seems like edge cases cannot happen, see WaveTrack::PasteWaveTrack
+   if (t0 == GetPlayStartTime())
+   {
+       ClearSequence(GetSequenceStartTime(), t0);
+       SetTrimLeft(other->GetTrimLeft());
+
+       auto copy = std::make_unique<WaveClip>(*other, mSequence->GetFactory(), true);
+       copy->ClearSequence(copy->GetPlayEndTime(), copy->GetSequenceEndTime());
+       newClip = std::move(copy);
+   }
+   else if (t0 == GetPlayEndTime())
+   {
+       ClearSequence(GetPlayEndTime(), GetSequenceEndTime());
+       SetTrimRight(other->GetTrimRight());
+
+       auto copy = std::make_unique<WaveClip>(*other, mSequence->GetFactory(), true);
+       copy->ClearSequence(copy->GetSequenceStartTime(), copy->GetPlayStartTime());
+       newClip = std::move(copy);
+   }
+   else
+   {
+       newClip = std::make_unique<WaveClip>(*other, mSequence->GetFactory(), true,
+           other->GetPlayStartTime(), other->GetPlayEndTime());
+   }
 
    if (clipNeedsResampling || clipNeedsNewFormat)
    {
-      newClip =
-         std::make_unique<WaveClip>(*other, mSequence->GetFactory(), true);
+      auto copy = std::make_unique<WaveClip>(*newClip.get(), mSequence->GetFactory(), true);
       if (clipNeedsResampling)
          // The other clip's rate is different from ours, so resample
-         newClip->Resample(mRate);
+          copy->Resample(mRate);
       if (clipNeedsNewFormat)
          // Force sample formats to match.
-         newClip->ConvertToSampleFormat(mSequence->GetSampleFormat());
-      pastedClip = newClip.get();
-   }
-   else
-   {
-      // No resampling or format change needed, just use original clip without making a copy
-      pastedClip = other;
+          copy->ConvertToSampleFormat(mSequence->GetSampleFormat());
+      newClip = std::move(copy);
    }
 
    // Paste cut lines contained in pasted clip
    WaveClipHolders newCutlines;
-   for (const auto &cutline: pastedClip->mCutLines)
+   for (const auto &cutline: newClip->mCutLines)
    {
-      newCutlines.push_back(
-         std::make_unique<WaveClip>
-            ( *cutline, mSequence->GetFactory(),
-              // Recursively copy cutlines of cutlines.  They don't need
-              // their offsets adjusted.
-              true));
-      newCutlines.back()->Offset(t0 - mOffset);
+      auto cutlineCopy = std::make_unique<WaveClip>(*cutline, mSequence->GetFactory(), 
+         // Recursively copy cutlines of cutlines.  They don't need
+         // their offsets adjusted.
+         true);
+      cutlineCopy->Offset(t0 - GetSequenceStartTime());
+      newCutlines.push_back(std::move(cutlineCopy));
    }
 
-   sampleCount s0;
-   TimeToSamplesClip(t0, &s0);
+   sampleCount s0 = TimeToSequenceSamples(t0);
 
    // Assume Strong-guarantee from Sequence::Paste
-   mSequence->Paste(s0, pastedClip->mSequence.get());
+   mSequence->Paste(s0, newClip->mSequence.get());
```

Comment:
There were three cases above:  pasting exactly at the play start time; exactly at play end; or other.

I think that in the first of these cases, the two lines above are incorrect.  Because ClearSequence() has removed some intial segment of samples, you need to paste at sample 0.

And I think you need to reassign mSequenceStartTime correctly in this case.

Am I mistaken?

Reply:
If I understand you correctly, then I think that the answer to the first question is: 
`sampleCount s0 = TimeToSequenceSamples(t0);`
`s0` should be 0 for given `t0` after `ClearSequence(GetSequenceStartTime(), t0);`
But I doubt that pasting in front of the clip is possible at all, taking into account implementation of the `WithinPlayRegion`